src/multirust-cli/self_update.rs

```diff
@@ -0,0 +1,1000 @@
+        while entry.th32ProcessID != this_pid {
+            let success = Process32Next(snapshot, &mut entry);
+            if success == 0 {
+                let err = io::Error::last_os_error();
+                return Err(Error::WindowsUninstallMadness(err));
+            }
+        }
+
+        let parent_id = entry.th32ParentProcessID;
+
+        // Get a handle to the parent process
+        let parent = OpenProcess(SYNCHRONIZE, 0, parent_id);
+        if parent == INVALID_HANDLE_VALUE {
+            let err = io::Error::last_os_error();
+            return Err(Error::WindowsUninstallMadness(err));
+        }
+
+        defer! {{ let _ = CloseHandle(parent); }}
+
+        // Wait for our parent to exit
+        let res = WaitForSingleObject(parent, 1000);
+
+        if res != WAIT_OBJECT_0 {
+            let err = io::Error::last_os_error();
+            return Err(Error::WindowsUninstallMadness(err));
+        }
+    }
+
+    Ok(())
+}
+
+#[cfg(unix)]
+pub fn complete_windows_install() -> Result<()> {
+    panic!("stop doing that")
+}
+
+#[derive(PartialEq)]
+enum PathUpdateMethod {
+    RcFile(PathBuf),
+    Windows,
+}
+
+/// Decide which rcfiles we're going to update, so we
+/// can tell the user before they confirm.
+fn get_add_path_methods() -> Vec<PathUpdateMethod> {
+    if cfg!(windows) {
+        return vec![PathUpdateMethod::Windows];
+    }
+
+    let bashrc = env::home_dir().map(|p| p.join(".bashrc"));
+    let zshrc = env::home_dir().map(|p| p.join(".zshrc"));
+    let kshrc = env::home_dir().map(|p| p.join(".kshrc"));
+
+    let rcfiles = vec![bashrc, zshrc, kshrc];
+    let existing_rcfiles = rcfiles.into_iter()
+        .filter_map(|f|f)
+        .filter(|f| f.exists());
+
+    existing_rcfiles.map(|f| PathUpdateMethod::RcFile(f)).collect()
+}
+
+fn shell_export_string() -> Result<String> {
+    let path = format!("{}/bin", try!(canonical_cargo_home()));
+    Ok(format!(r#"export PATH="{}:$PATH""#, path))
+}
+
+#[cfg(unix)]
+fn do_add_to_path(methods: &[PathUpdateMethod]) -> Result<()> {
+
+    for method in methods {
+        if let PathUpdateMethod::RcFile(ref rcpath) = *method {
+            let file = try!(utils::read_file("rcfile", rcpath));
+            let ref addition = format!("\n{}", try!(shell_export_string()));
+            if !file.contains(addition) {
+                try!(utils::append_file("rcfile", rcpath, addition));
+            }
+        } else {
+            unreachable!()
+        }
+    }
+
+    Ok(())
+}
+
+#[cfg(windows)]
+fn do_add_to_path(methods: &[PathUpdateMethod]) -> Result<()> {
+    assert!(methods.len() == 1 && methods[0] == PathUpdateMethod::Windows);
+
+    use winreg::RegKey;
+    use winapi::*;
+    use user32::*;
+    use std::ptr;
+
+    let root = RegKey::predef(HKEY_CURRENT_USER);
+    let environment = try!(root.open_subkey_with_flags("Environment", KEY_READ | KEY_WRITE)
+                           .map_err(|_| Error::PermissionDenied));
```

Comment:
I may just be getting lost and confused in the error handling here, but it seems weird that the original error is getting thrown away and it's being overwritten with a "permission denied" error?